### PR TITLE
Enable PPO on Intel XPU using a tiny model

### PR DESCRIPTION
**Pull request by songhappy**
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature

Please link to any issues this PR addresses.
https://jira.devtools.intel.com/browse/IPB-2914

#### Changelog
Added a configuration file of running PPO on Intel PVC 48G. 

#### Test plan

- [x] manually run any new or modified recipes with sufficient proof of correctness
- [x] I did not change any public API

### CODE REVIEW

**Comment by SalmanMohammadi** [comment]
Hi @songhappy. Thanks for opening this. It's great to see support in torchtune for different hardware and smaller models, but I'm hesitant to land a config for an `xpu` device as we won't be able to test and maintain it going forward. We generally test and ship our configs on `cuda` and allow users to configure it on their own.

**Review by ebsmothers — COMMENTED** [review]
Hi @songhappy thanks for the PR! For this case I think it would make sense to host this config outside of torchtune core. If I understand correctly the main difference is that this will run on XPU + the usage of a tiny Llama model, right? We don't really have any configs with tiny llama (as far as I know) so I think it would be a bit strange to add for just this one case. Let me know if this makes sense to you, thanks!

**Comment by songhappy** [comment]
Thanks for reviewing it. Actually, TinyLlama  is not really tiny, it's 1B, and we have a couple of 1B or 0.5B model configurations in here. For example, https://github.com/pytorch/torchtune/blob/main/recipes/configs/llama3_2/1B_full_single_device.yaml, https://github.com/pytorch/torchtune/blob/main/recipes/configs/qwen2/0.5B_full_single_device.yaml.

PPO uses way more memory than many other finetuning algorithms, given limited  resources on single device, please consider one configuration using 1B model for PPO. 

In terms of default device of CUDA, I can test it on CUDA and update this PR. 
Many thanks again. @SalmanMohammadi @ebsmothers

**Comment by ebsmothers** [comment]
@songhappy thanks, that makes sense. Given it's already 1B, can we use the Llama 3.2 1B model instead? I think this should give better results anyways

**Comment by SalmanMohammadi** [comment]
> @songhappy thanks, that makes sense. Given it's already 1B, can we use the Llama 3.2 1B model instead? I think this should give better results anyways

We don't have a classifier builder for 3.2. We can land #2356 soonish to enable this.

**Comment by songhappy** [comment]
> Hi @songhappy. Sorry for the substantial delay here. Now that #2356 has landed, would you be up for updating this config to use one of the more modern tiny models, e.g. Llama 3.2 1B like Evan suggested, or one of the Qwen2.5 models? If you could share the outputs of a successful run with an updated config I'm happy to land this.

sounds good, will look for different data to run a llama 3.2 1B or qwen2.5

**Comment by songhappy** [comment]
I do not find a finetuned llama3.2 1B classifier or finetuned Qianwen classifier as reward and value model. How about we still use this tiny llama of 1B?

**Review by joecummings — COMMENTED** [review]
Hi @songhappy - I don't actually think we want to land a specific config for XPU models, sorry for the delay in a decision here. 

I know we landed support for XCCL and some updated docs yesterday. If the goal for this PR was to make it clear that XPU devices can handle small configs for good RL things, we can add another note to our docs or we'd be happy to help write a blog that conveys this point.

What do you think?

**Comment by songhappy** [comment]
> Hi @songhappy - I don't actually think we want to land a specific config for XPU models, sorry for the delay in a decision here.
> 
> I know we landed support for XCCL and some updated docs yesterday. If the goal for this PR was to make it clear that XPU devices can handle small configs for good RL things, we can add another note to our docs or we'd be happy to help write a blog that conveys this point.
> 
> What do you think?

The goal for this recipe is to make a relatively small model for PPO recipe since this algorithm uses way more memory than, it does not have to be on XPU. If we can land this one with this recipe, I will double evaluate it on A100, and make cuda as default.

**Comment by joecummings** [comment]
> > Hi @songhappy - I don't actually think we want to land a specific config for XPU models, sorry for the delay in a decision here.
> > I know we landed support for XCCL and some updated docs yesterday. If the goal for this PR was to make it clear that XPU devices can handle small configs for good RL things, we can add another note to our docs or we'd be happy to help write a blog that conveys this point.
> > What do you think?
> 
> The goal for this recipe is to make a relatively small model for PPO recipe since this algorithm uses way more memory than, it does not have to be on XPU. If we can land this one with this recipe, I will double evaluate it on A100, and make cuda as default.

Understood! Once you fix the linting, we can get this landed then.

**Comment by songhappy** [comment]
> > > Hi @songhappy - I don't actually think we want to land a specific config for XPU models, sorry for the delay in a decision here.
> > > I know we landed support for XCCL and some updated docs yesterday. If the goal for this PR was to make it clear that XPU devices can handle small configs for good RL things, we can add another note to our docs or we'd be happy to help write a blog that conveys this point.
> > > What do you think?
> > 
> > 
> > The goal for this recipe is to make a relatively small model for PPO recipe since this algorithm uses way more memory than, it does not have to be on XPU. If we can land this one with this recipe, I will double evaluate it on A100, and make cuda as default.
> 
> Understood! Once you fix the linting, we can get this landed then.

All ready! thank you very much!

**Comment by SalmanMohammadi** [comment]
@songhappy I had a quick look for some small reward models. Could you see if any of these are suitable?

https://huggingface.co/Ray2333/GRM-Llama3.2-3B-rewardmodel-ft
https://huggingface.co/models?pipeline_tag=text-classification&sort=downloads&search=reward

If the 3B Llama3.2 model is a suitable size, would you be up for updating the config to use that, and share evidence of a successful run using it?

**Comment by songhappy** [comment]
> @songhappy I had a quick look for some small reward models. Could you see if any of these are suitable?
> 
> https://huggingface.co/Ray2333/GRM-Llama3.2-3B-rewardmodel-ft https://huggingface.co/models?pipeline_tag=text-classification&sort=downloads&search=reward
> 
> If the 3B Llama3.2 model is a suitable size, would you be up for updating the config to use that, and share evidence of a successful run using it?

1B model takes 20G, 3B model will take around 60G in memory, A100 does not have that much capacity. I think it's still worth to have this recipe of 1B PPO, even it's llama2.  

Here is what I collected on A100 for a 1B model. 
```
 nvidia-smi -l 1 --query-gpu=memory.used,memory.free,memory.total --format=csv
memory.used [MiB], memory.free [MiB], memory.total [MiB]
9003 MiB, 31324 MiB, 40960 MiB
9003 MiB, 31324 MiB, 40960 MiB
10051 MiB, 30276 MiB, 40960 MiB
10653 MiB, 29674 MiB, 40960 MiB
10663 MiB, 29664 MiB, 40960 MiB
18441 MiB, 21886 MiB, 40960 MiB
18441 MiB, 21886 MiB, 40960 MiB
18701 MiB, 21626 MiB, 40960 MiB
18701 MiB, 21626 MiB, 40960 MiB
18815 MiB, 21512 MiB, 40960 MiB
18815 MiB, 21512 MiB, 40960 MiB
19075 MiB, 21252 MiB, 40960 MiB
19075 MiB, 21252 MiB, 40960 MiB

```

**Comment by ebsmothers** [comment]
Thanks for your patience on this one @songhappy! I pushed a couple small changes (just minor cleanups), once CI finishes I think we can merge this.

**Review by ebsmothers — APPROVED** [review]

File: recipes/configs/llama2/1B_full_ppo_low_memory_single_device.yaml [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+# parameter for penalising generations shorter than `min_response_length`
+min_response_length: 18
+# parameter for penalising generations without a stop token
+penalise_no_eos: True
+# scalar penalty to apply when penalising
+reward_penalty: -3
+
+# tokens to consider as "end of sequence" tokens
+stop_token_ids: [
+  29889
```
Comment (SalmanMohammadi) [inline recipes/configs/llama2/1B_full_ppo_low_memory_single_device.yaml:154]:
```suggestion
stop_token_ids: [
  2,
  29889
```

File: recipes/configs/llama2/1B_full_ppo_low_memory_single_device.yaml [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+
+# Training arguments
+batch_size: 16
+num_steps: 1000
+ppo_epochs: 1
+ppo_batch_size: 16
+gradient_accumulation_steps: 1  # Use to increase effective batch size
+
+# Memory management and performance
+compile: True  # torch.compile the model + loss, True increases speed + decreases memory
```
Comment (SalmanMohammadi) [inline recipes/configs/llama2/1B_full_ppo_low_memory_single_device.yaml:126]:
```suggestion
compile: False  # torch.compile the model + loss, True increases speed + decreases memory
```

File: recipes/configs/llama2/1B_full_ppo_low_memory_single_device.yaml [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+# This config has been tested on an Intel PVC 48GB.
+# This config uses hyperparameters based on small set of experiments and information
+# available from existing implementations.
+#
+# This config assumes that you've run the following command before launching
+# this run:
+#   tune download smohammadi/tinyllama_rm_sentiment_1b --output-dir /tmp/tinyllama_rm_sentiment_1b/
+#   tune download TinyLlama/TinyLlama_v1.1 --output-dir /tmp/TinyLlama_v1.1/
+#
+# You'll also need to ensure that {output_dir} exists beforehand, as checkpoints for policy and value models are saved in sub-folders.
```
Comment (SalmanMohammadi) [inline recipes/configs/llama2/1B_full_ppo_low_memory_single_device.yaml:13]:
```suggestion
```

**Review by SalmanMohammadi — APPROVED** [review]
Thank you for your patience on this one @songhappy
